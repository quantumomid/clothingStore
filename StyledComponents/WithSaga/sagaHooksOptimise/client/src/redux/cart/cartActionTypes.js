export const TOGGLE_CART_HIDDEN = "TOGGLE_CART_HIDDEN";
export const ADD_ITEM = "ADD_ITEM";
export const CLEAR_ITEM_FROM_CART = "CLEAR_ITEM_FROM_CART";
export const REMOVE_ITEM = "REMOVE_ITEM";
export const CLEAR_CART = "CLEAR_CART";
export const SET_CART_FROM_FIREBASE = "SET_CART_FROM_FIREBASE";
export const UPDATE_CART_IN_FIREBASE = "UPDATE_CART_IN_FIREBASE";