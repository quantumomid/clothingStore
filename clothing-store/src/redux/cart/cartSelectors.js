import { createSelector } from "reselect";

//input selector  - gets the whole state and returns a slice of it (i.e. one layer deep usually)
const selectCart = state => state.cart;

export const selectCartItems = createSelector(
    [selectCart],
    cart => cart.cartItems
);

export const selectCartItemsCount = createSelector(
    [selectCartItems],
    cartItems => cartItems.reduce((accumulatedQuantity, cartItem) => accumulatedQuantity+cartItem.quantity, 0)
);