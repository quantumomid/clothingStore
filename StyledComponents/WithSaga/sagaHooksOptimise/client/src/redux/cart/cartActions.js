import { ADD_ITEM, CLEAR_CART, CLEAR_ITEM_FROM_CART, REMOVE_ITEM, TOGGLE_CART_HIDDEN, UPDATE_CART_IN_FIREBASE, SET_CART_FROM_FIREBASE } from "./cartActionTypes";

export const toggleCartHidden = () => {
    return { type: TOGGLE_CART_HIDDEN }
}

export const addItem = (item) => ({
    type: ADD_ITEM,
    payload: item
})

export const clearItemFromCart = item => ({
    type: CLEAR_ITEM_FROM_CART,
    payload: item
});

export const removeItem = item => ({
    type: REMOVE_ITEM,
    payload: item
});

export const clearCart = () => ({
    type: CLEAR_CART
});

// Needs no argument because updating in FIREBASE not our front end i.e. not redux
export const updateCartItemsInFirebase = () => ({
    type: UPDATE_CART_IN_FIREBASE
});

// Now need to pass cart items as we are setting the front end i.e. redux cart items state!
export const setCartFromFirebase = cartItems => ({
    type: SET_CART_FROM_FIREBASE,
    payload: cartItems
});