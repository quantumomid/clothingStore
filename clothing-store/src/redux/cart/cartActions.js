import { ADD_ITEM, TOGGLE_CART_HIDDEN } from "./cartActionTypes";

export const toggleCartHidden = () => {
    return { type: TOGGLE_CART_HIDDEN }
}

export const addItem = (item) => ({
    type: ADD_ITEM,
    payload: item
})