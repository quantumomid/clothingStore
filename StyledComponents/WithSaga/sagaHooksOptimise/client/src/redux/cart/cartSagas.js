import { takeLatest, put, all, call, select } from "redux-saga/effects";
import { SIGN_OUT_SUCCESS, SIGN_IN_SUCCESS } from "../user/userActionTypes";
import { clearCart, setCartFromFirebase, updateCartItemsInFirebase } from "./cartActions";
import { getUserCartRef } from "../../firebase/firebase.utils";
import { selectCurrentUser } from '../user/userSelectors';
import { selectCartItems } from './cartSelectors';
import { ADD_ITEM, CLEAR_ITEM_FROM_CART, REMOVE_ITEM } from "./cartActionTypes";

export function* clearCartOnSignOut(){
    yield put(clearCart());
}

// This is an example of reusing saga from another slice i.e.
// userSagas here to also perform some logic i.e. clear cart!
export function* onSignOutSuccess(){
    yield takeLatest(SIGN_OUT_SUCCESS, clearCartOnSignOut);
}

// This is run with respect to the SIGN_IN_SUCCESS action type 
// which comes with the user payload (see user actions JS file)
export function* checkCartFromFirebase({ payload: user }) {
    console.log("user in checkCartFromFirebase Saga", user);
    const cartRef = yield getUserCartRef(user.id);
    console.log("user Ref back from getUserCartRef in checkCartFromFirebase Saga", cartRef);
    const cartSnapshot = yield cartRef.get();
    yield put(setCartFromFirebase(cartSnapshot.data().cartItems));
}

// Our whole operation begins open successful sign-in > then only 
// will we check the cart items for the user in firebase
export function* onUserSignIn() {
    yield takeLatest(SIGN_IN_SUCCESS, checkCartFromFirebase);
}

// update the cart items in firebase as the user adds, removes or 
// clears their cart in the front end to keep firebase cart items 
// and redux ones in sync!
export function* updateCartInFirebase() {
    const currentUser = yield select(selectCurrentUser);
    if (currentUser) {
      try {
        const cartRef = yield getUserCartRef(currentUser.id);
        const cartItems = yield select(selectCartItems);
        yield cartRef.update({ cartItems });
        yield put(updateCartItemsInFirebase());
      } catch (error) {
        console.log(error);
      }
    }
}
  
// We also need to track whenever our cart state in redux changes 
// so we can update our carts in firebase ACCORDINGLY!
export function* onCartChange() {
    yield takeLatest(
      [
        ADD_ITEM,
        REMOVE_ITEM,
        CLEAR_ITEM_FROM_CART
      ],
      updateCartInFirebase
    );
}

export function* cartSagas(){
    yield all([call(onSignOutSuccess), call(onUserSignIn), call(onCartChange)]);
}