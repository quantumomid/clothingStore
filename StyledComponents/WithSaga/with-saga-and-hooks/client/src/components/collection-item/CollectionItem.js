import { useDispatch } from "react-redux";
import { addItem } from "../../redux/cart/cartActions";
import { AddCartButton, BackgroundImage, CollectionFooterContainer, CollectionItemContainer, NameContainer, PriceContainer } from "./collectionItemStyles";


function CollectionItem({ item }){
    const dispatch = useDispatch();
    const { name, price, imageUrl } = item
    return (
        <CollectionItemContainer>
            <BackgroundImage className='image' imageUrl={imageUrl} />
            <CollectionFooterContainer>
                <NameContainer>{name}</ NameContainer>
                <PriceContainer>{price}</ PriceContainer>
            </ CollectionFooterContainer>
            <AddCartButton onClick={() => dispatch(addItem(item))} inverted>Add to cart</AddCartButton>
        </ CollectionItemContainer>
    )
}

export default CollectionItem;