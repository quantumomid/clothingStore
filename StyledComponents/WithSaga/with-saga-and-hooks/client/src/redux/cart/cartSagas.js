import { takeLatest, put, all, call } from "redux-saga/effects";
import { SIGN_OUT_SUCCESS } from "../user/userActionTypes";
import { clearCart } from "./cartActions";

export function* clearCartOnSignOut(){
    yield put(clearCart());
}

// This is an example of reusing saga from another slice i.e.
// userSagas here to also perform some logic i.e. clear cart!
export function* onSignOutSuccess(){
    yield takeLatest(SIGN_OUT_SUCCESS, clearCartOnSignOut);
}

export function* cartSagas(){
    yield all([call(onSignOutSuccess)]);
}