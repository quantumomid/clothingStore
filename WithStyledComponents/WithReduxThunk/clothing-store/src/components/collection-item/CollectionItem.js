import { connect } from "react-redux";
import { addItem } from "../../redux/cart/cartActions";
import { AddCartButton, BackgroundImage, CollectionFooterContainer, CollectionItemContainer, NameContainer, PriceContainer } from "./collectionItemStyles";


function CollectionItem({ item, addItem }){
    const { name, price, imageUrl } = item
    return (
        <CollectionItemContainer>
            <BackgroundImage className='image' imageUrl={imageUrl} />
            <CollectionFooterContainer>
                <NameContainer>{name}</ NameContainer>
                <PriceContainer>{price}</ PriceContainer>
            </ CollectionFooterContainer>
            <AddCartButton onClick={() => addItem(item)} inverted>Add to cart</AddCartButton>
        </ CollectionItemContainer>
    )
}

const mapDispatchToProps = dispatch => {
    return {
        addItem: item => dispatch(addItem(item))
    }
}

export default connect(null, mapDispatchToProps)(CollectionItem);