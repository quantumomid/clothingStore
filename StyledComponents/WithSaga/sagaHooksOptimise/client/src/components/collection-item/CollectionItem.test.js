import { shallow } from "enzyme";
import * as redux from 'react-redux';
import { addItem } from "../../redux/cart/cartActions";
import CollectionItem from "./CollectionItem";

describe("Collection Item Component", () => {
    let wrapper;
    let spyOnUseDispatch;
    let mockDispatch;
    const mockItem =    {
        id: 3,
        name: 'Brown Cowboy',
        imageUrl: 'https://i.ibb.co/QdJwgmp/brown-cowboy.png',
      //   imageUrl: "/images/shop-img/hats/brown-cowboy.png",
        price: 35
      }

    beforeEach(() => {

        // Mock useDispatch hook
        spyOnUseDispatch = jest.spyOn(redux, 'useDispatch');
        // Mock dispatch function returned from useDispatch
        mockDispatch = jest.fn();
        spyOnUseDispatch.mockReturnValue(mockDispatch);

        wrapper = shallow(<CollectionItem item={mockItem} />);
    });
  
    afterEach(() => {
        jest.restoreAllMocks();
    });

    it("Collection Item Item component should be rendered", () => {
          expect(wrapper).toMatchSnapshot();
    })

    it('Should render the price of the Collection item as the text', () => {
        const price = parseInt(wrapper.find('PriceContainer').text());
        expect(price).toBe(35);
    });

    it('Should call dispatch when add button is clicked', () => {
        wrapper.find('AddCartButton').simulate('click');
        expect(mockDispatch).toHaveBeenCalled();
    });

    it('Should call dispatch with addItem method when add button is clicked', () => {
    wrapper.find('AddCartButton').simulate('click');
    expect(mockDispatch).toHaveBeenCalledWith(addItem(mockItem));
    });

    it('Should render the name of the Collection item as the text', () => {
        const name = wrapper.find('NameContainer').text();
        expect(name).toBe("Brown Cowboy");
    });

    it('should render imageUrl as a prop on BackgroundImage', () => {
        expect(wrapper.find('BackgroundImage').prop('imageUrl')).toBe("https://i.ibb.co/QdJwgmp/brown-cowboy.png");
    });
    
})
