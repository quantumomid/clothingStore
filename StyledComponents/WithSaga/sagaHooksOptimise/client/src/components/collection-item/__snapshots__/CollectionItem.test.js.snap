// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Collection Item Component Collection Item Item component should be rendered 1`] = `ShallowWrapper {}`;
